samples/read_bam.c

```diff
@@ -0,0 +1,169 @@
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE
+
+*/
+
+/* The pupose of this code is to demontrate the library apis and need proper error handling and optimization */
+
+#include <getopt.h>
+#include <unistd.h>
+#include <htslib/sam.h>
+
+/// print_usage - print the demo_usage
+/** @param fp pointer to the file / terminal to which demo_usage to be dumped
+returns nothing
+*/
+static void print_usage(FILE *fp)
+{
+    fprintf(fp, "Usage: read_bam [-n position] [-i infile] [<infile>]\n\
+Shows the alignment in given position\n");
+}
+
+/// main_demo - start of the demo
+/** @param argc - count of arguments
+ *  @param argv - pointer to array of arguments
+returns 1 on failure 0 on success
+*/
+int main(int argc, char *argv[])
+{
+    static const struct option lopts[] = {
+        {NULL, 0, NULL, 0}
+    };
+
+    const char *inname = NULL;
+    int c = 0, ret = EXIT_FAILURE, idx = 0;
+    sam_hdr_t *in_samhdr = NULL;
+    samFile *infile = NULL;
+
+    int ret_r = 0, i = 0, cnt = 0;
+    bam1_t *bamdata = NULL;
+    uint8_t *data = NULL;
+    uint32_t *cigar = NULL;
+
+
+    //read_bam [-n position] [-i infile] [<infile>]
+    while ((c = getopt_long(argc, argv, "i:n:", lopts, NULL)) >= 0)
+    {
+        switch (c)
+        {
+        case 'i':                           //in file name
+            inname = optarg;
+            break;
+        case 'n':                           //position
+            idx = atoi(optarg);
+            break;
+        case '?':                           //unknown options found!
+            print_usage(stderr);
+            goto end;
+            break;
+        }
+    }
+
+    if (argc == 1 && isatty(STDIN_FILENO)) {
+        print_usage(stdout);
+        ret = EXIT_SUCCESS;
+        goto end;
+    }
+
+    if (!inname && ((argc - optind) == 1))
+        inname = argv[optind];
+    if (!inname)
+        inname = "-";
+
+    bamdata = bam_init1();                  //bam data initialization
+    if (!bamdata)
+    {
+        fprintf(stderr, "Failed to allocate data memory!\n");
+        goto end;
+    }
+
+    //open input files
+    infile = sam_open(inname, "r");
+    if (!infile) {
+        fprintf(stderr, "Could not open %s\n", inname);
+        goto end;
+    }
+    //read header
+    in_samhdr = sam_hdr_read(infile);
+    if (!in_samhdr)
+    {
+        fprintf(stderr, "Failed to read header from file!\n");
+        goto end;
+    }
+
+    while (0 <= (ret_r = sam_read1(infile, in_samhdr, bamdata)))
+    {
+        if (idx != cnt++)
+            continue;
+
+        //QNAME FLAG RNAME POS MAPQ CIGAR RNEXT PNEXT TLEN SEQ QUAL [TAG:TYPE:VALUE]…
+        fprintf(stdout, "%s", bam_get_qname(bamdata));                                  //get the query name using the macro
```

Comment:
While printing out the contents is a good way of demonstrating the API (or lack of API via bam structure access), I think it is a mistake to print it up in a deliberate SAM looking format incase people mistakenly believe this is an example of BAM to SAM conversion.

Rather I think something like:

```
Read 1:
     Name\tFoo
     Flag\t0x41
     Chr\t1
     Pos\t1000
     ...
Read 2:
    ...
```

It's very clearly not SAM format, but its own thing. Maybe document it as a tool to read a BAM and report or dump the contents to stdout.


Also no need for `fprintf(stdout,...)`.  That's just `printf()`.

Reply:
Updated.